userspace/libsinsp/runc.h

```diff
@@ -0,0 +1,47 @@
+/*
+Copyright (C) 2013-2019 Draios Inc dba Sysdig.
+
+This file is part of sysdig.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+
+#pragma once
+
+#include <string>
+
+class sinsp_threadinfo;
+
+namespace libsinsp {
+namespace runc {
+
+/// runc-based runtimes (Docker, containerd, CRI-O, probably others) use the same two cgroup layouts
+/// with slight variations:
+/// - non-systemd layout uses cgroups ending with .../<prefix><container id>
+/// - systemd layout uses .../<prefix><container id>.scope
+/// where <container id> is always 64 hex digits (we report the first 12 as the container id).
+/// For non-systemd only CRI-O seems to use /crio-<container id>, while for systemd layout
+/// while all known container engines use a prefix like "docker-", "crio-" or "containerd-cri-".
+/// We can encode all these variants with a simple list of (prefix, suffix) pairs
+/// (the last one must be a pair of null pointers to mark the end of the array)
+struct cgroup_layout {
+	const char* prefix;
+	const char* suffix;
+};
+
+/// If any of the cgroups of the thread in `tinfo` matches the `layout`, set `container_id` to the found id
```

Comment:
what if multiple cgroups match the layout?

I assume we're assuming that impossible, and if that's the case ,we should perhaps note the assumption

Reply:
Then we take the first matching one (AFAICT, the iteration order matches contents of /proc/<pid>/cgroups). If the cgroups are inconsistent (matching different containers), there's no single right answer we can choose. What do you propose?

Reply:
so long as we're okay with that, it's fine, but we should at least be explicit in that "if it matches multiple, we'll only return the first" or something